src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,303 @@
+#include "rgw_aws.h"
+#include <string>
+#include <iostream>
+#include <utility>
+#include "common/dout.h"
+#include <aws/core/Aws.h>
+#include <aws/sns/SNSClient.h>
+#include <aws/lambda/LambdaClient.h>
+#include <aws/lambda/model/InvokeRequest.h>
+#include <boost/lockfree/queue.hpp>
+#include <aws/core/utils/ARN.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/sns/model/PublishRequest.h>
+
+#define dout_subsys ceph_subsys_rgw
+
+namespace rgw::aws {
+  static const int STATUS_OK = 1;
+  static const int STATUS_FULL = 2;
+  static const int STATUS_MANAGER_STOPPED = 3;
+  static const int STATUS_CLIENT_CLOSED = 4;
+
+  struct client_id_t {
+    const std::string accessKey;
+    const std::string accessSecret;
+    const std::string arn;
+    const bool useSSL;
+
+    client_id_t(const std::string &accessKey, const std::string &accessSecret,
+                const std::string &arn, bool useSsl) : accessKey(accessKey),
+                                                      accessSecret(accessSecret),
+                                                      arn(arn), useSSL(useSsl) {}
+
+    bool operator==(const client_id_t &other) const {
+      return accessKey == other.accessKey &&
+             accessSecret == other.accessSecret &&
+             arn == other.arn && useSSL == other.useSSL;
+    }
+
+    struct hasher {
+      std::size_t operator()(const client_id_t &k) const {
+        return ((std::hash<std::string>()(k.accessKey)
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+  class AsyncCallerContext: public Aws::Client::AsyncCallerContext{
+  public:
+    reply_callback_t cb;
+    AsyncCallerContext(reply_callback_t cb):cb(cb){}
+  };
+
+  void invoke_callback(const Aws::Lambda::LambdaClient* client,
+                              const Aws::Lambda::Model::InvokeRequest& request,
+                              const Aws::Lambda::Model::InvokeOutcome & outcome,
+                              const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context){
+
+    auto derived = std::static_pointer_cast<const AsyncCallerContext>(context);
+    if (outcome.IsSuccess()) {
+      std::cout << "Invoke function: " << request.GetFunctionName() << "with body " << request.GetBody() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(STATUS_OK);
+      }
+    }
+    else {
+      const auto& error = outcome.GetError();
+      std::cout << "ERROR: " << error.GetExceptionName() << ": "
+                << error.GetMessage() << std::endl;
+      if(derived && derived->cb) {
+        derived->cb(-1);
```

Comment:
is there an actual error code we can pass to cb() ? e.g.
negative `GetStatusCode()`